### ISSUE TITLE
Flash attention and flash decoding principles

### ISSUE THREAD

**Issue by RonanKMcGovern** [ml-explore/mlx#129]
Are there plans to add flash attention and also flash decoding to allow for improved performance for long context?

**Comment by awni** [comment]
We'd love to have this. Our first priority is quantization but when we have the bandwidth we can look into adding Flash attention.  (Note PRs are welcome)

**Comment by BuildBackBuehler** [comment]
> We'd love to have this. Our first priority is quantization but when we have the bandwidth we can look into adding Flash attention. (Note PRs are welcome)

I'd messaged the maintainer of [this project](https://github.com/philipturner/metal-flash-attention) a few days ago because it seemed like he's dedicated to it and I saw he wanted to implement it in 1 or 2 other projects. But in case you can't get ahold of 'em, you have the link ¯\_(ツ)_/¯

**Comment by ivanfioravanti** [comment]
This would be amazing! So that we can have integration in the amazing axolotl!

**Comment by chigkim** [comment]
Sorry for a newbie question, but is Flash Attention specifically tied to a library? If there's a flash attention for Metal, could it be used by both Torch with MPS support and MLX?

A lot of recent models use flash-attn on pip. If we have something that works for Torch, Transformers, and MLX, that would be truly amazing!

**Comment by joseph777111** [comment]
https://github.com/philipturner/metal-flash-attention 

From a very knowledgeable friend, pertaining to this project vs mlx.

philipturner/metal-flash-attention, only implements single headed attention. Modern LLMs need multi headed attention. In any case, MLX's `mlx.core.fast.scaled_dot_product_attention` does implement flash attention. I've read the kernel src and verified it. Although IDK if the mlx devs are aware of it (there are a couple of open bugs on their tracker about it), because it was a 3rd party contrib. 😄

**Comment by awni** [comment]
> MLX's mlx.core.fast.scaled_dot_product_attention does implement flash attention. I've read the kernel src and verified it. 

There is a fused attention kernel.. but it's not used by default because it's still slower than the explicit implementation (e.g. two matmuls with a softmax etc).

We working on faster fused attention for both inference (single query) and prompt / training. Eventually that will be the default for `scaled_dot_product_attention`.

**Comment by yuchaoran2011** [comment]
Just saw this issue. I'd like to mention my experience using the latest mx.fast.scaled_dot_product_attention implementation. On my M3 Max, when training a GPT-2-sized model locally, I didn't observe noticeable training efficiency differences between the fast version and explicit attention computation. Most likely it's because my GPU is already at full utilization during training and the HBM tax that Flash Attention solves doesn't quite apply in Apple Silicon architecture. I'll be interested in other users' experience.
